MSC3765: Rich text in room topics (#3765)
matrix-org/matrix-spec-proposals: proposals/3765-rich-room-topics.md

Drawing from extensible events as described in [MSC1767],
`m.room.topic` is formally deprecated and replaced with a new `m.topic`
event type. The latter contains a new content block `m.topic` which wraps
an `m.text` content block that allows representing the room topic in
different mime types.

``` json5
{
  "type": "m.topic",
  "state_key": "",
  "content": {
    "m.topic": {
      "m.text": [{
          "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
          "mimetype": "text/html",
          "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }]
    }
  },
  ...
}
```

Details of how `m.text` works may be found in [MSC1767] and are not
repeated here.

The wrapping `m.topic` content block is similar to `m.caption` for file
uploads as defined in [MSC3551]. It avoids clients accidentally rendering
the topic state event as a room message.

In order to prevent formatting abuse in room topics, clients are
encouraged to treat the first two lines as the shorthand topic and the
remainder as additional information. Specifically, this means that
things like headings and enumerations should be ignored (or formatted

> as regular text) unless they occur after the second line.

Reviewer: This seems fairly arbitrary -- would having a separate field that defines the "short" topic make sense, or just leave it to an implementation detail?

Author: I believe this originally stemmed from the fact that Element Web was truncating the room topic shown on top of the timeline after the first two lines. These days it doesn't appear to display the topic on top of the timeline at all anymore though.

I suppose a short topic would still be desirable for e.g. room directories or space hierarchies. So a separate field sounds like a good idea. I guess at this point it could also be separate events though, e.g. `m.topic` for the short topic (which is how it's usually used today) and `m.details` (or similar) for an extended multiline description?

Author: Addressed in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/commits/f915cc4663c153f18c77b5fe05081eafc4e9d37c.

A change to `/_matrix/client/v3/createRoom` is not necessary. The
endpoint has a plain text `topic` parameter but also allows to specify a

> full `m.topic` event in `initial_state`.

Reviewer: Does this imply homeservers should now create `m.topic` when creating rooms? Both events?

Author: I think this has been covered in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/files#r1731217524.

Room topics also occur as part of the `PublicRoomsChunk` object in the
responses of `/_matrix/client/v3/publicRooms` and
`/_matrix/client/v1/rooms/{roomId}/hierarchy`. The topic can be kept
plain text here because this data should commonly only be displayed to
users that are *not* a member of the room yet. These users will not have
the same need for rich room topics as users who are inside the room. If
no plain text topic exists, home servers should return an empty topic
string from these end points. Since this will inevitably lead to bad UX,
client implementations are encouraged to always include a plain text

> variant when sending `m.topic` events.

Reviewer: Does this mean homeservers should search for `m.topic`, get a `text/plain` MIME type from it, then fallback to `m.room.topic`?

Author: I think this has been covered in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/files#r1731217524.

## Transition

As this MSC replaces `m.room.topic` for an extensible alternative,
clients and servers are expected to treat `m.room.topic` as invalid in

> extensible event-supporting room versions.

Reviewer: Maybe this answers my questions from above -- it is unclear the new events only apply to extensible events rooms.

Author: Yes, you're right. I think the bit at the beginning of the proposal

> `m.room.topic` is formally deprecated and replaced with a new `m.topic` event type

is making it unclear. It's not actually a deprecation but rather mandating the exclusive use of `m.topic` in extensible events rooms. Since a new room version will be required, think we won't need any fallback logic to handle both events. Let me try and clarify that.

Author: Updated the text at the beginning.

`m.topic` which wraps an `m.text` content block that allows representing
the room topic in different mime types.

``` json5
{
  "type": "m.topic",
  "state_key": "",
  "content": {
    "m.topic": {
      "m.text": [{
          "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
          "mimetype": "text/html",
          "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }]
    }
  },
  ...
}
```

Details of how `m.text` works may be found in [MSC1767] and are not
repeated here.

The wrapping `m.topic` content block is similar to `m.caption` for file
uploads as defined in [MSC3551]. It avoids clients accidentally rendering
the topic state event as a room message.

In order to prevent formatting abuse in room topics, clients are
encouraged to limit the length of topics to at most two lines. Additionally,
clients should ignore things like headings and enumerations (or format them
as regular text). A future MSC may introduce a mechanism to capture extended
multiline details that are not suitable for room topics in a separate field
or event type.

> A change to `/_matrix/client/v3/createRoom` is not necessary. The

Reviewer 2: non-blocking: link to the spec, and maybe a bit of introduction. The paragraph kinda comes out of left field.

Author: Links fixed with https://github.com/matrix-org/matrix-spec-proposals/pull/4215/commits/c86770feded9c4bf97cd1acece908251f9461935. Intro (and further details) provided with https://github.com/matrix-org/matrix-spec-proposals/pull/4215/commits/ab7b646f4157acb9cd6adc6e212a0f45532c839f.

endpoint has a plain text `topic` parameter but also allows to specify a
full `m.topic` event in `initial_state`.

Room topics also occur as part of the `PublicRoomsChunk` object in the
responses of `/_matrix/client/v3/publicRooms` and

> `/_matrix/client/v1/rooms/{roomId}/hierarchy`. The topic can be kept

Reviewer 2: non-blocking: links to spec please

Author: Fixed with https://github.com/matrix-org/matrix-spec-proposals/pull/4215/commits/c86770feded9c4bf97cd1acece908251f9461935.

plain text here because this data should commonly only be displayed to
users that are *not* a member of the room yet. These users will not have
the same need for rich room topics as users who are inside the room. If
no plain text topic exists, home servers should return an empty topic
string from these end points. Since this will inevitably lead to bad UX,
client implementations are encouraged to always include a plain text
variant when sending `m.topic` events.

## Transition

> As this MSC replaces `m.room.topic` for an extensible alternative,
> clients and servers are expected to treat `m.room.topic` as invalid in
> extensible event-supporting room versions.

Reviewer 2: blocking for FCP finish (not start): 
```suggestion
As this MSC replaces `m.room.topic` for an extensible alternative,
clients and servers are expected to treat `m.room.topic` as invalid in
extensible event-supporting room versions. Similarly, `m.topic` cannot
be used in non-extensible-supporting room versions.
```

Author: I cannot commit the suggestion here because I don't have push access but I've added it in https://github.com/matrix-org/matrix-spec-proposals/pull/4215/commits/620cdd3d26a4ddc175cc48b95c60f3be04098937.

## Potential issues

None.

## Alternatives

The combination of `format` and `formatted_body` currently utilised to
enable HTML in `m.room.message` events could be generalised to
`m.topic` events. However, this would only allow for a single
format in addition to plain text and is a weaker form of reuse than
described in the introductory section of [MSC1767].

## Security considerations

Allowing HTML in room topics is subject to the same security
considerations that apply to HTML in room messages.

## Unstable prefix

> While this MSC is not considered stable, `m.topic` should be referred to
> as `org.matrix.msc3765.topic`.

Author: ```suggestion
While this MSC is not considered stable, `m.topic` should be referred to as
`org.matrix.msc3765.topic`. Note that extensible events and content blocks
might have their own prefixing requirements.

### Implementation considerations

Implementation authors should note that as a feature using the Extensible Events
system, usage of the *stable* event type `m.topic` in regular room versions is not
permitted. As of writing (August 2024), Extensible Events does not have a *stable*
room version which supports such events. Therefore, implementations will have to
use the *unstable* event type if they intend to support rich text in room topics in
existing room versions.

When Extensible Events as a system is released in a dedicated room version, clients
will be able to use the stable event type there. The unstable event type should
not be used in that dedicated room version.
```

> ## Unstable prefix
> 
> While this MSC is not considered stable, `m.topic` should be referred to
> as `org.matrix.msc3765.topic`.

Author: ```suggestion
## Other notes

Normally extensible events would only be permitted in a specific room version.
However, this proposal's event is permitted in any room version. The stable
event type should only be sent in a room version which supports extensible
events, however.

## Unstable prefix

While this MSC is not considered stable, `m.topic` should be referred to as
`org.matrix.msc3765.topic`. Note that extensible events and content blocks
might have their own prefixing requirements.

### Implementation considerations

Implementation authors should note that as a feature using the Extensible Events
system, usage of the *stable* event type `m.topic` in regular room versions is not
permitted. As of writing (August 2024), Extensible Events does not have a *stable*
room version which supports such events. Therefore, implementations will have to
use the *unstable* event type if they intend to support rich text in room topics in
existing room versions.

When Extensible Events as a system is released in a dedicated room version, clients
will be able to use the stable event type there. The unstable event type should
not be used in that dedicated room version.
```

Reviewer: > Normally extensible events would only be permitted in a specific room version. However, this proposal's event is permitted in any room version. The stable event type should only be sent in a room version which supports extensible events, however.

So in this case you'd expect both the unstable version and `m.room.topic`? I'm not quite sure what this means.

Author: Hm, good question. Forbidding `m.room.topic` in existing room versions feels wrong. Maybe this situation calls for the coexistence of both `org.matrix.msc3765.topic` and `m.room.topic` with rules for reading and writing the topic in a room:

* If `org.matrix.msc3765.topic` exists, use that. Otherwise fall back to `m.room.topic`.
* When writing `org.matrix.msc3765.topic`, also write `m.room.topic` with just the plain text topic.

Does that sound sensible?

Reviewer: I think it sounds sensible, but it is just a bit finicky / complicated. E.g.:

1. You have both `m.topic` and `m.room.topic` set from a "modern" client, nominally the values are the same.
2. Someone using an "old" client updates the topic, but only `m.room.topic` is updated.
3. Users now see two different topics depending on what client they're using.

(This is also a potential security issue of presenting different information to different users, although it wouldn't be hidden per-say, just a display issue.)

Author: Yeah, true. 😕

The current implementation in Element Web actually has that problem even though it embedded the rich topic into `m.room.topic` (which was the first version of this proposal).

I don't think we can really avoid that unless we restrict `m.topic` to a new room version that forbids `m.room.topic`. Maybe pointing out the risk for early implementers in the proposal would suffice?

Reviewer: I don't have a better idea. Would be interested to hear from others though.

Author: In reconsidering this, I think it may be better to revert to _not_ allowing the use of `m.topic` in current room versions. This was originally proposed in an attempt to mimic the approach taken for polls in [MSC3381](https://github.com/matrix-org/matrix-spec-proposals/pull/3381). However, polls introduce entirely new events and, hence, are not subject to the same compatibility and security problems that replacing `m.room.topic` causes. Instead of this, I think we should take the approach from [MSC1767](https://github.com/matrix-org/matrix-spec-proposals/pull/1767) and require a new room version for `m.topic`.

Therefore, I propose _not_ to apply the suggestion at the start of this thread. I don't believe this should prevent this proposal from going through FCP and, if successful, be merged without being added to the spec yet (which is how [MSC1767](https://github.com/matrix-org/matrix-spec-proposals/pull/1767) itself was treated, too).

Reviewer 3: What's wrong with embedding the rich topic into `m.room.topic` for old room versions? That seems like the best option, since presumably old clients updating the topic would then clear the rich topic rather than leaving an old rich topic laying around.

Author: Hm, that's an interesting idea.

So new clients would write `m.room.topic` with both `m.topic` and `topic` (being the copy of the plain text body in `m.topic`) in `content` while old clients would only write `topic` (effectively deleting `m.topic`). If both behave correctly, we'd never have a situation where the old and new topic mismatch.

In places where the server needs the topic, it could read it from the plain text body in `m.topic`, falling back to `topic` if the former doesn't exist. On room upgrades, the server would also have to read the other mimetypes in `m.topic`.

Feels like relatively small risk and overhead that could enable using rich topics before extensible events land?

Reviewer 2: I have not read the whole thread here, but +1 to cramming this into `m.room.topic` for old room versions.

Author: Implemented in https://github.com/matrix-org/matrix-spec-proposals/pull/4215/commits/de8e9a295a1d71fb218f628d3f3d2118d4432ae2.

## Problem

Topics are a central piece of room meta data and usually made easily
accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

Drawing from extensible events as described in [MSC1767], a new content
block `m.topic` is defined, which wraps an `m.text` content block that
allows representing the room topic in different mime types. In current
room versions, this content block is added to the content of [`m.room.topic`]
events as shown below[^1].

```json5
{
  "type": "m.room.topic",
  "state_key": "",
  "content": {
    "m.topic": {
      "m.text": [{
        "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
        "mimetype": "text/html",
        "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"

>       }]

Reviewer 4: Aren't these the wrong way round? According to MSC1767, the *first* supported representation should be used, so nobody will ever use the html representation.

Reviewer 2: (clients can also optimize and pick a preferred represenation, but indeed the preference should be to order them. I remain unconvinced that MSC1767 picked the right order here though, because everyone (including me) keeps putting plaintext first)

Reviewer 4: I've raised a concern about getting this addressed.

Author: ```suggestion
      "m.text": [ {
        "mimetype": "text/html",
        "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }, {
        "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }]
```

    },
    "topic": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
  },
  ...
}
```

Details of how `m.text` works may be found in [MSC1767] and are not

> repeated here.

Reviewer 4: I hadn't really expected `m.room.topic` events to be the way that the `m.text` stuff makes it into the spec, but I guess that's ok.

The wrapping `m.topic` content block is similar to `m.caption` for file
uploads as defined in [MSC3551]. It avoids clients accidentally rendering

> the topic state event as a room message.

Reviewer 4: > It avoids clients accidentally rendering the topic state event as a room message.

Is this actually necessary?

It seems to me that the reason for `m.caption` in MSC3551 is quite different to the reason for it here: in that case, `m.text` at the top level has a very different meaning to `m.text` within `m.caption`, whereas here there is no ambiguity.

This might be a good thing to do anyway, for consistency for example, but I'm unconvinced this paragraph captures a good reason for it.

Reviewer 5: I tripped over this too when reviewing it just now. My initial reaction was the same as richvdh's: state events don't get rendered as fallbacks, plus clients know to special-case topics already - plus the contents of the topic is semantically a top-level 'm.text' (unlike m.caption, which is effectively describing a nested event).  If topics actually required some more custom topic-specific datatype (e.g. `show_at_all_times: true` or something) then perhaps one might wrap it... but in practice, I can't think of any plausible customisation which couldn't/shouldn't be pulled in as another mix-in.

The only argument I can see for it is consistency with m.caption and other places where you might want to explicitly say "don't fall back to displaying m.text if you don't recognise the event type. even if it's a state event". Or possibly futureproofing for additional topic-specific keys we just haven't thought of yet.

If it were me, I'd probably go ahead without the `m.topic` wrapper and just have it as `m.text`.  But my strength of feeling here is about 6/10.

Reviewer 5: So i've gone and talked this through with @turt2live, and I think I understand the rationale better:
 * If this was a new state event type, we'd put the content in a wrapper to prevent it showing up in timelines (if MSC1767 showed m.text blocks for unknown state events, which in retrospect we think it should - after all, the whole point of MSC1767 is to show human representation of unknown events where appropriate, whether they are timeline or state events).
 * Semantically, the fallback which *would* show up in timelines is different to the actual content of the event. For instance, a hypothetical callback `content['m.text']` could be "The topic was set to 'Foo'", whereas the actual content of the topic would be `content['m.topic']['m.text'])`
 * We probably *should* save somewhere for additional fields on topics in general. the `show_always: true` example is actually a valid example of something we could conceivably want to put on topics, for instance.

Therefore i've gone full circle (sorry), and think the wrapper is the right approach after all. Sorry for burning your time unnecessarily on #4251 😔

To allay @richvdh's original concern and help everyone else who trips over this, we should probably hoist this justification in the MSC.

Reviewer 2: https://github.com/matrix-org/matrix-spec-proposals/pull/4252 is an MSC about the state event handling, fwiw.

Author: > Therefore i've gone full circle (sorry), and think the wrapper is the right approach after all. Sorry for burning your time unnecessarily on https://github.com/matrix-org/matrix-spec-proposals/pull/4251 😔

No worries at all. I've switched https://github.com/matrix-org/matrix-spec-proposals/pull/4251 around to properly explain why the extra content block is needed.

Reviewer 4: I've merged #4251 so that we can all see the latest version of the text. I'm not sure we're quite there yet, but I'll pose my questions as comments on the new text.

> It is recommended that clients always include a plain text variant when

Reviewer 4: I thought this was referring to the `topic` property initially.
```suggestion
It is recommended that clients always include a plain text variant within `m.text` when
```

Author: Good point. Could you apply this for me as well please?

> sending `m.topic` events. This prevents bad UX in situations where a plain

Reviewer 4: ```suggestion
sending `m.room.topic` events. This prevents bad UX in situations where a plain
```

Author: This looks good to me. Could you apply the suggestion in my stead? Thanks for catching.

text topic is sufficient such as the public rooms directory.

Additionally, clients should duplicate the plain text topic into the existing
`topic` field for backwards compatibility with clients that don't support
`m.topic` yet. This also helps prevent inconsistencies since such clients
are likely to delete the `m.topic` content block when updating `m.room.topic`
themselves.

In order to prevent formatting abuse in room topics, clients are
encouraged to limit the length of topics during both entry and display,
for instance, by capping the number of displayed lines. Additionally,
clients should ignore things like headings and enumerations (or format them
as regular text). A future MSC may introduce a mechanism to capture extended
multiline details that are not suitable for room topics in a separate field
or event type.

> On the server side, any logic that currently operates on the `topic` field is
> updated to use the `m.topic` content block instead.

Reviewer 4: Am I right in understanding that the next four paragraphs are examples of this change? If so, it would be good to make that a little clearer:

```suggestion
On the server side, any logic that currently operates on the `topic` field is
updated to use the `m.topic` content block instead. For example:
```

... and make the following paragraphs bullet points.

Author: It's actually meant to be a conclusive list. I hope I didn't miss anything. Maybe ending with a colon (without "For example") and making the next paragraphs bullets would be clearer?

Reviewer 4: yup, sgtm

Author: ```suggestion
On the server side, any logic that currently operates on the `topic` field is
updated to use the `m.topic` content block instead:

- In [`/_matrix/client/v3/createRoom`], the `topic` parameter should cause `m.room.topic`
  to be written with a `text/plain` mimetype in `m.topic`. If at the same time an
  `m.room.topic` event is supplied in `initial_state`, it is overwritten entirely.
  A future MSC may generalize the `topic` parameter to allow specifying other mime
  types without `initial_state`.
- In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]
  and their `POST` siblings, the `topic` response field should be read from the
  `text/plain` mimetype of `m.topic` if it exists or omitted otherwise.
  A plain text topic is sufficient here because this data is commonly
  only displayed to users that are *not* a member of the room yet. These
  users don't commonly have the same need for rich room topics as users
  who already reside in the room. A future MSC may update these endpoints
  to support rich text topics.
- The same logic is applied to [`/_matrix/client/v1/rooms/{roomId}/hierarchy`]
  and [`/_matrix/federation/v1/hierarchy/{roomId}`].
- In [server side search], the `room_events` category is expanded to search
  over the `m.text` content block of `m.room.topic` events.
```

Author: Have captured it into https://github.com/matrix-org/matrix-spec-proposals/pull/3765/files#r1890019768 while I'll need you to push the button on once more. Sorry for the hassle.

> In [`/_matrix/client/v3/createRoom`], the `topic` parameter causes `m.room.topic`
> to be written with a `text/plain` mimetype in `m.topic`. If at the same time an

Reviewer 4: I thought this was a statement of current behaviour at first.

```suggestion
In [`/_matrix/client/v3/createRoom`], the `topic` parameter should cause `m.room.topic`
to be written with a `text/plain` mimetype in `m.topic`. If at the same time an
```

Author: Makes sense. Could you apply this, too, please?

`m.room.topic` event is supplied in `initial_state`, it is overwritten entirely.
A future MSC may generalize the `topic` parameter to allow specifying other mime
types without `initial_state`.

In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]

> and their `POST` siblings, the `topic` response field is read from the

Reviewer 4: ```suggestion
and their `POST` siblings, the `topic` response field should be read from the
```

Author: Makes sense. Could you apply this, too, please?

> On the server side, any logic that currently operates on the `topic` field is
> updated to use the `m.topic` content block instead.
> 
> In [`/_matrix/client/v3/createRoom`], the `topic` parameter should cause `m.room.topic`
> to be written with a `text/plain` mimetype in `m.topic`. If at the same time an
> `m.room.topic` event is supplied in `initial_state`, it is overwritten entirely.
> A future MSC may generalize the `topic` parameter to allow specifying other mime
> types without `initial_state`.
> 
> In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]
> and their `POST` siblings, the `topic` response field should be read from the
> `text/plain` mimetype of `m.topic` if it exists or omitted otherwise.
> A plain text topic is sufficient here because this data is commonly
> only displayed to users that are *not* a member of the room yet. These
> users don't commonly have the same need for rich room topics as users
> who already reside in the room. A future MSC may update these endpoints
> to support rich text topics.
> 
> The same logic is applied to [`/_matrix/client/v1/rooms/{roomId}/hierarchy`]
> and [`/_matrix/federation/v1/hierarchy/{roomId}`].
> 
> In [server side search], the `room_events` category is expanded to search
> over the `m.text` content block of `m.room.topic` events.

Author: ```suggestion
On the server side, any logic that currently operates on the `topic` field is
updated to use the `m.topic` content block instead:

- In [`/_matrix/client/v3/createRoom`], the `topic` parameter should cause `m.room.topic`
  to be written with a `text/plain` mimetype in `m.topic`. If at the same time an
  `m.room.topic` event is supplied in `initial_state`, it is overwritten entirely.
  A future MSC may generalize the `topic` parameter to allow specifying other mime
  types without `initial_state`.
- In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]
  and their `POST` siblings, the `topic` response field should be read from the
  `text/plain` mimetype of `m.topic` if it exists or omitted otherwise.
  A plain text topic is sufficient here because this data is commonly
  only displayed to users that are *not* a member of the room yet. These
  users don't commonly have the same need for rich room topics as users
  who already reside in the room. A future MSC may update these endpoints
  to support rich text topics.
- The same logic is applied to [`/_matrix/client/v1/rooms/{roomId}/hierarchy`]
  and [`/_matrix/federation/v1/hierarchy/{roomId}`].
- In [server side search], the `room_events` category is expanded to search
  over the `m.text` content block of `m.room.topic` events.
```

## Potential issues

None.

## Alternatives

The combination of `format` and `formatted_body` currently utilised to
enable HTML in `m.room.message` events could be generalised to
`m.room.topic` events. However, this would only allow for a single
format in addition to plain text and is a weaker form of reuse than
described in the introductory section of [MSC1767].

## Security considerations

Allowing HTML in room topics is subject to the same security
considerations that apply to HTML in room messages. In particular,
topics are already included in the content that clients should [sanitise]
for unsafe HTML.

## Unstable prefix

While this MSC is not considered stable, `m.topic` should be referred to
as `org.matrix.msc3765.topic`. Note that extensible events and content

> blocks might have their own prefixing requirements.

Reviewer 4: > Note that extensible events and content blocks might have their own prefixing requirements.

I don't really know what this means, as it pertains to this MSC?

Author: This is stolen from [MSC3381](https://github.com/matrix-org/matrix-spec-proposals/pull/3381/files#diff-4ea5a1fe3badf80ded5be7b58d532db7b851a93f1b455ae7302e6f3f4d433a2aR383-R384). It was originally intended to make `m.text` into `org.matrix.msc1767.text`. However, I suppose we don't need this anymore because [MSC1767](https://github.com/matrix-org/matrix-spec-proposals/pull/1767) has since been accepted?

Reviewer 4: I see. Yes, I don't think we need this.

Author: ```suggestion
as `org.matrix.msc3765.topic`.
```

# MSC3765: Rich text in room topics

## Problem

Topics are a central piece of room meta data and usually made easily
accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

Drawing from extensible events as described in [MSC1767], a new content
block `m.topic` is defined, which wraps an `m.text` content block that
allows representing the room topic in different mime types. In current
room versions, this content block is added to the content of [`m.room.topic`]
events as shown below[^1].

```json5
{
  "type": "m.room.topic",
  "state_key": "",
  "content": {

>     "m.topic": {

Reviewer 6: Just for clarity here (since we've always failed to spec this behavior outright). Should `m.topic` be omitted if the topic was removed from the room, or empty array (or both because it amounts to the same thing?).

The implementation doesn't allow for this currently, and I'm just doing some cleanup.

Author: I don't really have a strong opinion on this at all but in an attempt to make the decision feel less arbitrary: `content["m.topic"]` will have to be optional because otherwise we'd make all existing topic events suddenly be incompatible. In other words, omitting `m.topic` will have to be possible. Since that already provides us _a_ way to remove the rich topic, we might as well say that it is _the_ way. Happy to hear other thoughts, however.

      "m.text": [{
        "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
        "mimetype": "text/html",
        "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }]
    },
    "topic": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
  },
  ...
}
```

> Details of how `m.text` works may be found in [MSC1767] and are not
> repeated here.

Author: Addendum to https://github.com/matrix-org/matrix-spec-proposals/pull/3765/files#r1926496551.

```suggestion
In line with [MSC1767], clients should render the first mime type in the
array that they understand. Further details of how `m.text` works may
be found in [MSC1767] and are not repeated here.
```

updated to use the `m.topic` content block instead:

- In [`/_matrix/client/v3/createRoom`], the `topic` parameter should cause `m.room.topic`
  to be written with a `text/plain` mimetype in `m.topic`. If at the same time an
  `m.room.topic` event is supplied in `initial_state`, it is overwritten entirely.
  A future MSC may generalize the `topic` parameter to allow specifying other mime
  types without `initial_state`.
- In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]
  and their `POST` siblings, the `topic` response field should be read from the
  `text/plain` mimetype of `m.topic` if it exists or omitted otherwise.
  A plain text topic is sufficient here because this data is commonly
  only displayed to users that are *not* a member of the room yet. These
  users don't commonly have the same need for rich room topics as users
  who already reside in the room. A future MSC may update these endpoints
  to support rich text topics.
- The same logic is applied to [`/_matrix/client/v1/rooms/{roomId}/hierarchy`]
  and [`/_matrix/federation/v1/hierarchy/{roomId}`].
- In [server side search], the `room_events` category is expanded to search

>   over the `m.text` content block of `m.room.topic` events.

Reviewer 4: @Johennes this bit doesn't seem to have been implemented in mainline synapse (unless I'm missing it): do we have a sample server-side implementation?

Author: Better late than never: https://github.com/element-hq/synapse/pull/18195

The advantage of having shunted the implementation is that it now can use stable identifiers directly. 😇

None.

## Alternatives

The combination of `format` and `formatted_body` currently utilised to
enable HTML in `m.room.message` events could be generalised to
`m.room.topic` events. However, this would only allow for a single
format in addition to plain text and is a weaker form of reuse than
described in the introductory section of [MSC1767].

## Security considerations

Allowing HTML in room topics is subject to the same security
considerations that apply to HTML in room messages. In particular,
topics are already included in the content that clients should [sanitise]
for unsafe HTML.

## Unstable prefix

While this MSC is not considered stable, `m.topic` should be referred to
as `org.matrix.msc3765.topic`. Note that extensible events and content
blocks might have their own prefixing requirements.

[^1]: A future MSC may discuss how to adopt the `m.topic` content block in
      new room versions which support extensible events.

[plain text]: https://spec.matrix.org/v1.12/client-server-api/#mroomtopic
[MSC1767]: https://github.com/matrix-org/matrix-spec-proposals/pull/1767

> [MSC3551]: https://github.com/matrix-org/matrix-spec-proposals/pull/3551

Author: ```suggestion
```

This won't be needed anymore when https://github.com/matrix-org/matrix-spec-proposals/pull/3765#discussion_r1923883332 is applied.

# MSC3765: Rich text in room topics

## Problem

Topics are a central piece of room meta data and usually made easily
accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

Drawing from extensible events as described in [MSC1767],
`m.room.topic` is formally deprecated and replaced with a new `m.topic`
event type. The latter contains an `m.markup` content block representing
the room topic in different mime types.

``` json5
{
    "type": "m.topic",
    "state_key": "",
    "content": {

>         "m.markup": [{

Reviewer 2: To avoid clients accidentally rendering the topic event as a room message (as funny as that would be), I'd suggest *also* introducing an `m.topic` content block, like `m.caption` for file uploads: https://github.com/matrix-org/matrix-spec-proposals/pull/3551

Reviewer 2: Note that under MSC1767's latest updates (made minutes ago as of writing), clients *shouldn't* be trying to treat state events as timeline events at all, but the MSC still doesn't prohibit a client from doing so if it really wants to.

Reviewer 2: (MSC1767 also changed `m.markup` to be `m.text` in the final hour)

Author: This will be addressed with https://github.com/matrix-org/matrix-spec-proposals/pull/4111.

            "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
        }],
    },
    ...
}
```

Details of how `m.markup` works may be found in [MSC1767] and are not
repeated here.

In order to prevent formatting abuse in room topics, clients are
encouraged to treat the first two lines as the shorthand topic and the
remainder as additional information. Specifically, this means that
things like headings and enumerations should be ignored (or formatted
as regular text) unless they occur after the second line.

While the content of `m.topic` is currently identical to `m.message`, a
dedicated event type allows the two to diverge in the future.

A change to `/_matrix/client/v3/createRoom` is not necessary. The
endpoint has a plain text `topic` parameter but also allows to specify a
full `m.topic` event in `initial_state`.

Room topics also occur as part of the `PublicRoomsChunk` object in the
responses of `/_matrix/client/v3/publicRooms` and
`/_matrix/client/v1/rooms/{roomId}/hierarchy`. The topic can be kept
plain text here because this data should commonly only be displayed to
users that are *not* a member of the room yet. These users will not have

> the same need for rich room topics as users who are inside the room.

Reviewer 2: what if the topic event only declares an html rendering? (`m.markup` allows for plaintext to be optional, for better or worse)

Author: This will be addressed with https://github.com/matrix-org/matrix-spec-proposals/pull/4112.

> ## Transition

Reviewer 2: A couple notes on the transition: the intention is clear, though I don't think we need a dedicated room version for topic events. We can just say "As this MSC replaces `m.room.topic` for an extensible alternative, clients and servers are expected to treat `m.room.topic` as invalid in extensible event-supporting room versions."

This suggestion would also override all the language about "greater than or equal to N". Note that room versions are non-linear and have no guaranteed ordering: they are effectively flagged off feature support instead.

Author: This will be addressed with https://github.com/matrix-org/matrix-spec-proposals/pull/4112.

[`/rooms/{roomId}/upgrade`]. If the new room version is greater or
equal to N, an existing `m.room.topic` event in the old room

``` json5
{
    "type": "m.room.topic",
    "state_key": "",
    "content": {
        "topic": "All about pizza"
    },
    ...
}
```

should be migrated to an `m.topic` event with a single plain-text markup
in the new room

``` json5
{
    "type": "m.topic",
    "state_key": "",
    "content": {
        "m.markup": [{
            "mimetype": "text/plain",
            "body": "All about pizza"
        }],
    },
    ...
}
```

## Potential issues

None.

## Alternatives

The combination of `format` and `formatted_body` currently utilised to
enable HTML in `m.room.message` events could be generalised to
`m.topic` events. However, this would only allow for a single
format in addition to plain text and is a weaker form of reuse than
described in the introductory section of [MSC1767].

## Security considerations

Allowing HTML in room topics is subject to the same security
considerations that apply to HTML in room messages.

## Unstable prefix

While this MSC is not considered stable, `m.topic` should be referred to

> as `org.matrix.msc3765.topic`.

Reviewer 2: @jplatte says:

The current implementation in the JS SDK adds `"org.matrix.msc3765.topic"` as a content field to the existing `m.room.topic` event, but this MSC documents a new `m.topic` event `type`. Is the JS SDK implementation just outdated?

options.

## Proposal

Drawing from extensible events as described in [MSC1767], `m.room.topic`
is prohibited in room versions that support extensible events and replaced
with a new `m.topic` event type. The latter contains a new content block
`m.topic` which wraps an `m.text` content block that allows representing
the room topic in different mime types.

``` json5
{
  "type": "m.topic",
  "state_key": "",
  "content": {
    "m.topic": {
      "m.text": [{
          "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
          "mimetype": "text/html",
          "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }]
    }
  },
  ...
}
```

Details of how `m.text` works may be found in [MSC1767] and are not
repeated here.

The wrapping `m.topic` content block is similar to `m.caption` for file
uploads as defined in [MSC3551]. It avoids clients accidentally rendering
the topic state event as a room message.

It is recommended that clients always include a plain text variant when
sending `m.topic` events. This prevents bad UX in situations where a plain

> text topic is sufficient such as the public rooms directory.

Reviewer 7: Not a blocker but this feels like an odd example as I don't see what the public room directory would necessarily choose not to use rich text. Text-only clients maybe?

Author: Blocking on https://github.com/matrix-org/matrix-spec-proposals/pull/3765/files#r1812651036 where we're discussing whether we'll include all mime types everywhere. If so, this would make this whole paragraph obsolete.

Author: If agreed in the discussion mentioned in my last comment to postpone rich topics in the room directory and other endpoints to a future MSC. So I believe this comment here can be resolved.

In order to prevent formatting abuse in room topics, clients are

> encouraged to limit the length of topics to at most two lines. Additionally,

Reviewer 8: At most two HTML lines? Or if one crafts a very long plain text topic that would be wrapped into more than two lines that should be limited as well?

I understand where this comes from but in its current reading the recommendation is a bit confusing.

Author: What I meant is two displayed lines – regardless of whether the lines resulted from actual newlines in the topic or from a long topic being wrapped onto multiple lines for display.

Does that make sense and if so, do you think "limit the length of topics to at most two _displayed_ lines" would make this clearer?

Reviewer: It might make sense to just say that clients should be sure to limit it, e.g. by limiting the number of displayed lines.

Author: This was addressed in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/commits/1b4682e43a6ba6fa44af1cfbd3694fe512302f8c.

> encouraged to limit the length of topics to at most two lines. Additionally,

Reviewer 7: At both entry and display time presumably?

Author: This was addressed in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/commits/1b4682e43a6ba6fa44af1cfbd3694fe512302f8c.

> In order to prevent formatting abuse in room topics, clients are
> encouraged to limit the length of topics to at most two lines. Additionally,
> clients should ignore things like headings and enumerations (or format them
> as regular text). A future MSC may introduce a mechanism to capture extended
> multiline details that are not suitable for room topics in a separate field
> or event type.

Author: ```suggestion
In order to prevent formatting abuse in room topics, clients are
encouraged to limit the length of topics during both entry and display,
for instance, by capping the number of displayed lines. Additionally,
clients should ignore things like headings and enumerations (or format them
as regular text). A future MSC may introduce a mechanism to capture extended
multiline details that are not suitable for room topics in a separate field
or event type.
```

This is to address the comments of @KitsuneRal, @clokep and @dbkr. If this is alright, could one of you please accept the suggestion in my stead?

Reviewer: I think that sounds fine -- does this work for you @KitsuneRal & @dbkr?

Reviewer: I accepted it, but shout if this doesn't help

On the server side, any logic that currently operates on `m.room.topic` is
updated to use `m.topic` instead.

In [`/_matrix/client/v3/createRoom`], the `topic` parameter causes `m.topic`
to be written with a `text/plain` mimetype. If an `m.topic` event is supplied
in `initial_state`, the `topic` parameter overwrites its `text/plain` mimetype

> but retains any other mimetypes.

Reviewer 7: This feels like a footgunny way to end up with a text and html part that don't match. I would expect specifying a topic event in initial_state would take precedence completely rather than require the server to edit parts of the event.

Author: Hm, that sounds sensible but then we'd have to swap the order of steps 6 and 7 in https://spec.matrix.org/v1.12/client-server-api/#post_matrixclientv3createroom depending on the room version?

Reviewer 7: Ah yep, good point. It feels like the right answer to me.

Reviewer 9: If we want to preserve the current behaviour, then it seems to me like it would be better to have the `topic` parameter overwrite the whole `m.topic` event.

Author: This was addressed in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/commits/d42f0e36866c397932f969c0e9a1050499b9fab1.

> In [`/_matrix/client/v3/createRoom`], the `topic` parameter causes `m.topic`
> to be written with a `text/plain` mimetype. If an `m.topic` event is supplied
> in `initial_state`, the `topic` parameter overwrites its `text/plain` mimetype
> but retains any other mimetypes.

Author: > If we want to preserve the current behaviour, then it seems to me like it would be better to have the `topic` parameter overwrite the whole `m.topic` event.

I think I like that best as it's quite easy to do and may still be improved by a future MSC generalizing (or possibly even eliminating) the `topic` parameter.

@dbkr / @uhoreg how does this sound?

```suggestion
In [`/_matrix/client/v3/createRoom`], the `topic` parameter causes `m.topic`
to be written with a `text/plain` mimetype. If at the same time an `m.topic`
event is supplied in `initial_state`, it is overwritten entirely. A future MSC
may generalize the `topic` parameter to allow specifying other mime types
without `initial_state`.
```

Reviewer 9: lgtm.  @dbkr any objections?

In [`GET /_matrix/client/v3/publicRooms`], [`GET /_matrix/federation/v1/publicRooms`]
and their `POST` siblings, the `topic` response field is read from the
`text/plain` mimetype of `m.topic` if it exists or omitted otherwise.
A plain text topic is sufficient here because this data is commonly
only displayed to users that are *not* a member of the room yet. These
users don't have the same need for rich room topics as users who already

> reside in the room.

Reviewer 7: Hmm, I see why you might say this but I'm not sure I agree, and this is going to mean the public room list topics are full of weird looking markdown syntax. I'm inclined to think that if we want topics to be HTML, they ought to be HTML everywhere.

Author: We could maybe achieve that by cutting a new version of the endpoint and including all mimetypes in the `topic` field (similar to `m.text` in the event itself)?

```json5
{
  "chunk": [
    {
      "avatar_url": "mxc://bleecker.street/CHEDDARandBRIE",
      "name": "CHEESE",
      "topic": [{
          "body": "All about **cheese**"
      }, {
          "mimetype": "text/html",
          "body": "All about <b>cheese</b>"
      }],
      ...
    }
  ],
  ...
}
```

This would also have to be applied to the SS endpoints as well as the space hierarchy endpoints. I had not considered this so far because it felt slightly invasive. I agree that it's a bit odd to not have rich topics everywhere though.

Reviewer 7: Yeah, this would feel quite reasonable. It could be left for a separate MSC though, so this MSC ends up being the same technically, just with a, "public rooms list doesn't support html topics yet" rather than, "we've decided that public room list doesn't support html topics".

Author: This was addressed in https://github.com/matrix-org/matrix-spec-proposals/pull/3765/commits/d42f0e36866c397932f969c0e9a1050499b9fab1.

> A plain text topic is sufficient here because this data is commonly
> only displayed to users that are *not* a member of the room yet. These
> users don't have the same need for rich room topics as users who already
> reside in the room.

Author: ```suggestion
A plain text topic is sufficient here because this data is commonly
only displayed to users that are *not* a member of the room yet. These
users don't commonly have the same need for rich room topics as users
who already reside in the room. A future MSC may update these endpoints
to support rich text topics.
```

@dbkr / @uhoreg does this sound better and if so, could one of you apply it in my stead?

Reviewer 9: lgtm

The same logic is applied to [`/_matrix/client/v1/rooms/{roomId}/hierarchy`]
and [`/_matrix/federation/v1/hierarchy/{roomId}`].

In [server side search], the `room_events` category is expanded to search

> over the `text/plain` mimetype in `m.topic`.

Reviewer 7: This feels overly restrictive I think - I don't see any reason servers shouldn't use other fields to search if they wish.

Author: Yes, I think you're right. If this sounds ok, could you accept the suggestion in my stead?

```suggestion
over the `m.text` content block of `m.topic` events.
```

# MSC3765: Rich text in room topics

## Problem

Topics are a central piece of room meta data and usually made easily
accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

> Drawing from extensible events as described in [MSC1767], `m.room.topic`
> is prohibited in room versions that support extensible events and replaced
> with a new `m.topic` event type. The latter contains a new content block
> `m.topic` which wraps an `m.text` content block that allows representing
> the room topic in different mime types.

Reviewer 4: I read this and thought "well, since there are no current room versions that support extensible events, we're not going to be able to use this anywhere".

Having read through a bit more and come across the "Transition" and "Notes" sections, I realise that's not really the case.

What's more: I am worried that any behaviour defined here for some mythical extensible-events-supporting room version is going to be forgotten by the time we actually come to define such a room version; and then we will find ourselves very confused about what is or is not specced. I would strongly argue that we should limit this proposal to the room versions we have today, with some non-normative *suggestions* about how how things might look in future room versions.

In other words: let's rephrase this a bit.

```suggestion
Drawing from extensible events as described in [MSC1767], a new content block
`m.topic` is defined, which wraps an `m.text` content block that allows representing
the room topic in different mime types.

In current room versions, this content block is added to the content of 
[`m.room.topic`](https://spec.matrix.org/v1.12/client-server-api/#mroomtopic) events as shown below.
See also (link to separate section) which discusses possible adoption in future room versions which
support extensible events.
```

Author: It sounds like this would mean rebasing this proposal onto the current spec (which doesn't know extensible events) and then creating another proposal to adapt it to the extensible events room version later?

I'd be ok with that but I feel like I'm being bounced back and forth a little here. @turt2live this would probably benefit from your input.

@richvdh I guess you might want to stick up a concern?

Reviewer 2: Over the last month I've thought about this a bit more, and I'm very supportive of bypassing the Extensible Events stuff for the time being. At the time when we tied this MSC to extensible events, there was hope that the whole system would land by about this time in the year, but that both hasn't happened and has no team working on it. 

There's also been vague discussion about whether a room version is even correct for extensible events, which may cause the MSC to slide back to in-review. I don't know what kind of impact this will have on already-accepted MSCs which rely on extensible events, but suffice to say they are definitely forgotten.

TLDR: Let's go the non-normative suggestions route, and land this (finally).

Author: Thanks a lot. I have opened https://github.com/matrix-org/matrix-spec-proposals/pull/4240 to address this.

Reviewer 2: (considering resolved - please reopen or open a new thread if required)

``` json5
{
  "type": "m.topic",
  "state_key": "",
  "content": {
    "m.topic": {
      "m.text": [{
          "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
      }, {
          "mimetype": "text/html",
          "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
      }]
    }
  },
  ...
}

> ```

Reviewer 4: replace with example of m.room.topic event

> # MSC3765: Rich text in room topics

Reviewer 2: @alphapapa says:

> On one hand, I can see some elegance in repurposing room topics for general-purpose, long-term room reference information. OTOH, it seems like overloading the purpose of topics with what, in other systems, would go in "pinned" topics or messages, or a wiki, etc.
> 
> So IMHO I would consider implementing support for pinned messages/events before overloading topics like this. It would seem relatively straightforward for a room's state to have a list of pinned events, which could be sent in initial sync by the server or be retrieved manually by clients. Clients could then display these pinned events in a room's timeline view, optionally hiding them, compressing them, etc. And the pinned events could be edited by room moderators using existing event-editing tools. (Forgive me if there's already a proposal for something like that.)

Reviewer 2: @Johennes replies:

> Interesting idea. Pinned events seem to already [exist](https://spec.matrix.org/v1.2/client-server-api/#mroompinned_events). However, in their current form, these are not fit to be used for what you describe because, depending on room settings, users joining the room _after_ the events were sent could be unable to see them.

accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

To enrich `m.room.topic` events, we build upon extensible events as
defined in [MSC1767] and define a new `m.topic` event in `content`.
The latter contains a list of renderings in the same format that
[MSC1767] uses for `m.message` events.

```json5
{
    "type": "m.room.topic",
    "state_key": "",
    "content": {
        "m.topic": [{
            "mimetype": "text/html", // optional, default text/plain
            "body": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>"
        }, {
            "mimetype": "text/plain",
            "body": "All about **pizza** | [Recipes](https://recipes.pizza.net)"
        }],
    },
    ...
}
```

While the content of `m.topic` is currently identical to `m.message`, a

> dedicated event type allows the two to diverge in the future.

Reviewer 10: can you give some examples of why it may be useful to allow m.topic to diverge from m.message, or is the complexity there just for theoretical flexibility?

Author: Answered in the comment above. I think the separation may not be needed after all.

responses of `/_matrix/client/v3/publicRooms` and
`/_matrix/client/v1/rooms/{roomId}/hierarchy`. The topic can be kept
plain text here because this data should commonly only be displayed to
users that are *not* a member of the room yet. These users will not have
the same need for rich room topics as users who are inside the room.

## Transition

Similar to [MSC1767] a time-constrained transition period is proposed.
Upon being included in a released version of the specification, the
following happens:

-   The `topic` field in the content of `m.room.topic` events is
    deprecated
-   Clients continue to include `topic` in outgoing events as a fallback
-   Clients prefer the new `m.topic` format in events which include it
-   A 1 year timer begins for clients to implement the above conditions
    -   This can be shortened if the ecosystem adopts the format sooner
    -   After the (potentially shortened) timer, an MSC is introduced to
        remove the deprecated `topic` field. Once accepted under the
        relevant process, clients stop including the field in outgoing
        events.

## Potential issues

None.

> ## Alternatives

Reviewer 2: @emorrp1 says:

> Hi, thanks for working on this (real hyperlinks!), I love the sample screenshot.
> 
> Can you explain why an Alternative using `m.relates_to` is not possible here, is it because of the same join issue above? It just seems odd to me to have a topic-specific message format, rather than "linking" to an existing message (which itself could be an extensible event).
> 
> I guess a related question is can you give some examples of why it may be useful to allow `m.topic` to diverge from `m.message`, or is the complexity there just for theoretical flexibility?

Author: @emorrp1: Happy to hear you're excited about links in topics, too. 😅

If you're referring to a _timeline_ message event, I think the same join issue mentioned above would apply here, too, because any timeline event is subject to the room's history rules. In order to have the event available for anyone who joins the room, it has to be a _state_ event.

We could probably post such topic messages as state events themselves but that's then not much different from allowing the message formats on the existing `m.room.topic` event (which is what this proposal does).

Maybe this is an indicator that `m.topic` and `m.message` should, in fact, not be allowed to diverge? Keeping them separate was more of a theoretical thought and I can't think of a concrete use case where we'd need this separation right now.

Author: This is obsolete now that we introduced a dedicated `m.topic` content block to avoid rendering topic events as room messages accidentally.

# MSC3765: Rich text in room topics

> ## Problem

Reviewer 5: This is great :) I wonder if we should put some guidelines to stop people shoving H1 tags in the topic or other stuff which is going to make for an ugly UX though... (or recommend that clients should ignore certain tags to preserve sane aesthetics? We don't want to reinvent myspace)

Author: How about we suggest clients to ignore things like headings and enumerations on the first two lines but allow them afterwards? This way they could continue to render the beginning of the topic above the timeline while still allowing for arbitrarily rich formatting in the remainder of the topic which is only displayed in a dedicated screen / overlay.

Here's an example of a very long topic in Element Web:

![Screenshot 2022-04-04 at 19 58 59](https://user-images.githubusercontent.com/1137962/161603712-d31bce9f-98f5-4adf-abee-5afd82eef672.png)

Author: I have worked in a section to cover this.

Topics are a central piece of room meta data and usually made easily
accessible to room members in clients. As a result, room administrators
often extend the use of topics to collect helpful peripheral information
that is related to the room’s purpose. Most commonly these are links to
external resources. At the moment, topics are limited to [plain text]
which, depending on the number and length of URLs and other content,
easily gets inconvenient to consume and calls for richer text formatting
options.

## Proposal

To enrich `m.room.topic` events, we build upon extensible events as
defined in [MSC1767] and define a new `m.topic` event in `content`.
The latter contains a list of renderings in the same format that
[MSC1767] uses for `m.message` events.

> ``` json

Reviewer 2: To stop github complaining about the `...`
```suggestion
```json5
```

Author: TIL!

{
    "type": "m.room.topic",
    "state_key": "",

>     "content": {

Reviewer 5: My immediate reaction on seeing the long-form `m.topic: [{...}, {...}]` body was "wow, this looks more complicated than i'd expect".  This is why I added `m.text` and `m.html` to MSC1767, so that the example here would actually be:

```json5
"m.topic": {
   "m.text": "All about **pizza** | [Recipes](https://recipes.pizza.net)",
   "m.html": "All about <b>pizza</b> | <a href=\"https://recipes.pizza.net\">Recipes</a>",
}
```

...which just feels much nicer to develop against and debug against, for the super-common case of a normal rich message - as well as using much less room.

Is the intention to support the short-form as well as long-term event shape?

Reviewer 2: MSC1767 is starting to question a lot of the benefit for shortcuts and may ultimately remove them. There's also been massive interest in not supporting the shortcuts for types like this, as it gets confusing if it's meant to be an array or an object with types.

I have to admit I'm very much in favour of dropping the shortcuts at this stage given they seem to be providing more confusion than cleanliness. We should take this discussion over to MSC1767 though.

Reviewer 11: +1 for short form. more concise, easier to read, take less room and carry the same information. in fact `m.text` alone with markdown links will do most of the time.

Author: Resolving this as this MSC directly follows what [MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767) does in terms of shorthand notation. The latest form of [MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767) does not include shortcuts but if they shall be reintroduced there, the intention would be that they'd work in the context of this MSC as well.

>         "m.topic": [{

Reviewer 12: Wouldn't this be entirely equivalent to `m.message`?

Author: Yes, true. I kept the two apart since, unlike messages, topics are normally not rendered in the timeline so maybe there could be a usecase for divergence in the future.

Reviewer 12: That's fine, I'm just noting it, it might be useful to note both those things ("it's similar to `m.message`", and "it can diverge from it in the future") in the proposal text.

Author: Good point. I added a phrase capturing this below the example.

    deprecated
-   Clients continue to include `topic` in outgoing events as a fallback
-   Clients prefer the new `m.topic` format in events which include it
-   A 1 year timer begins for clients to implement the above conditions
    -   This can be shortened if the ecosystem adopts the format sooner
    -   After the (potentially shortened) timer, an MSC is introduced to
        remove the deprecated `topic` field. Once accepted under the
        relevant process, clients stop including the field in outgoing
        events.

## Potential issues

None.

## Alternatives

The combination of `format` and `formatted_body` currently utilised to
enable HTML in `m.room.message` events could be generalised to
`m.room.topic` events. However, this would only allow for a single
format in addition to plain text and is a weaker form of reuse as
described in the introductory section of [MSC1767].

## Security considerations

Allowing HTML in room topics is subject to the same security
considerations that apply to HTML in room messages.

## Unstable prefix

While this MSC is not considered stable, implementations should apply
the behaviours of this MSC on top of room version 9 as

> org.matrix.msc3765.

Reviewer 2: This doesn't require a room version - just the point under this paragraph is fine (maybe keeping the words "while this MSC is not considered stable, `m.topic` should be ..."

Author: Sorry, this was poor copy/pasting on my side. Fixed now.

The wrapping `m.topic` content block is similar to `m.caption` for file
uploads as defined in [MSC3551]. It avoids clients accidentally rendering

> the topic as a room message. Note that while [MSC1767] had explicitly
> excluded state events from being treated as extensible, this is being
> changed with [MSC4252]. The extra content block, therefore, allows putting

Reviewer 4: Non-blocking comment:

I think we could do with clarifying *why* the `m.topic` block avoids rendering as a room message: this was the main thing I was missing when I opened the other comment thread:

```suggestion
the topic as a room message. ([MSC1767] specifies that unknown events with
an `m.text` content block should be rendered as a regular room message, and
while [MSC1767] had explicitly excluded state events from being treated as 
extensible, this is being changed with [MSC4252].) The extra content block, therefore, allows putting
```

a fallback representation that is actually designated for the timeline

> into a separate `content['m.text']` field. Lastly, the `m.topic` content

Reviewer 4: Non-blocking comment:

AFAICT, there are only two arguments here, so "Lastly" reads strangely. 

```suggestion
into a separate `content['m.text']` field. In addition, the `m.topic` content
```

Author: I counted avoiding rendering as a fallback and consciously supplying a fallback in a separate field as two. But "In addition" works fine either way.

I'm also fine with your other two suggestions above but will need somebody else to apply them again. I'm sorry for the mess this creates but I don't know what else to do about it. 😔

block also serves as a good place for additional fields to be added by

> other MSCs in the future.

Reviewer 4: Non-blocking comment:

I don't entirely follow why we need an `m.topic` block as a place to put extra fields (unless we think that we're going to end up with topics inside *other* event types?) but maybe that's my lack of familiarity with extensible events.

@ara4n as the person that proposed this (I think?) can you clarify?
